rasa/nlu/classifiers/keyword_intent_classifier.py

```diff
@@ -1,32 +1,140 @@
-from typing import Any, Optional, Text
+import os
+import logging
+import typing
+import re
+from typing import Any, Dict, Optional, Text
 
+from rasa.nlu import utils
 from rasa.nlu.components import Component
 from rasa.nlu.training_data import Message
 
+logger = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
+    from rasa.nlu.config import RasaNLUModelConfig
+    from rasa.nlu.training_data import TrainingData
+    from rasa.nlu.model import Metadata
+    from rasa.nlu.training_data import Message
+
 
 class KeywordIntentClassifier(Component):
+    """Intent classifier using simple keyword matching.
+
+
+    The classifier takes a list of keywords and associated intents as an input.
+    A input sentence is checked for the keywords and the intent is returned.
+
+    """
 
     provides = ["intent"]
 
-    his = ["hello", "hi", "hey"]
+    defaults = {"case_sensitive": True}
 
-    byes = ["bye", "goodbye"]
+    def __init__(
+        self,
+        component_config: Optional[Dict[Text, Any]] = None,
+        intent_keyword_map: Optional[Dict] = None,
+    ):
 
-    def process(self, message: Message, **kwargs: Any) -> None:
+        super(KeywordIntentClassifier, self).__init__(component_config)
 
-        intent = {"name": self.parse(message.text), "confidence": 1.0}
-        message.set("intent", intent, add_to_output=True)
+        if intent_keyword_map is None:
+            self.intent_keyword_map = {}
+        else:
+            self.intent_keyword_map = intent_keyword_map
 
-    def parse(self, text: Text) -> Optional[Text]:
+    def train(
+        self,
+        training_data: "TrainingData",
+        cfg: Optional["RasaNLUModelConfig"] = None,
+        **kwargs: Any
+    ) -> None:
 
-        _text = text.lower()
+        for intent in training_data.intents:
+            self.intent_keyword_map[intent] = [
+                ex.text
+                for ex in training_data.training_examples
+                if ex.get("intent") == intent
+            ]
+        self._validate_keyword_map()
 
-        def is_present(x):
-            return x in _text
+    def _validate_keyword_map(self):
+        re_flags = 0 if self.component_config["case_sensitive"] else re.IGNORECASE
+        for i, (intent1, ex1s) in enumerate(self.intent_keyword_map.items()):
+            for j, (intent2, ex2s) in enumerate(self.intent_keyword_map.items()):
+                if j != i:
+                    for ex1 in ex1s:
+                        for ex2 in ex2s:
+                            if re.search(r"\b" + ex1 + r"\b", ex2, flags=re_flags):
```

Comment:
oh no, why so many `for`?

Reply:
In order to print a clear warning, i.e. one that shows which keywords conflict with which, we need to check the examples of every intent against the each example of every other intent. 
So I think implicitely there will always be 4 loops and I can't think of a nicer way to code this. Any ideas?

Reply:
Can we check it on the fly in the for above inverting `intent_keyword_map`: `ex.text` will be keys and `intent` is a value, if this text already exists delete both and add a warning. Similar as `MemoizationPolicy` is done

Reply:
great idea! now it's just two loops

Reply:
that's not what I meant. I suggested to do it during training every time you add a new intent, text pair